### ISSUE TITLE
Parallel valgrind errors

### ISSUE THREAD

**Issue by lindsayad** [libMesh/libmesh#1857]
It is possible that this is a red herring, but so far googles of "Unaddressable bytes mpiwrap false positives" have yielded no results. If I run `examples/vector_fe/ex3` with `LD_PRELOAD=/path/to/libmpiwrap mpirun -np 2 valgrind ./example-oprof` where `mpirun` is provided by the MOOSE package `mpich`, then I get these types of traces:
```
==25250== Unaddressable byte(s) found during client check request                                                                                                                                                                             
==25250==    at 0x4E415E0: check_mem_is_defined_untyped (libmpiwrap.c:953)                                                                                                                                                                    
==25250==    by 0x4E421DA: walk_type (libmpiwrap.c:691)                                                                                                                                                                                       
==25250==    by 0x4E49CE5: walk_type_array (libmpiwrap.c:924)                                                                                                                                                                                 
==25250==    by 0x4E49CE5: check_mem_is_defined (libmpiwrap.c:995)                                                                                                                                                                            
==25250==    by 0x4E49CE5: PMPI_Allreduce (libmpiwrap.c:2094)                                                                                                                                                                                 
==25250==    by 0x58D8E4E: max<unsigned int> (parallel_implementation.h:538)                                                                                                                                                                  
==25250==    by 0x58D8E4E: libMesh::MeshTools::n_levels(libMesh::MeshBase const&) (mesh_tools.C:648)                                                                                                                                          
==25250==    by 0x5932784: libMesh::UnstructuredMesh::find_neighbors(bool, bool) (unstructured_mesh.C:398)                                                                                                                                    
==25250==    by 0x57E61E7: libMesh::MeshBase::prepare_for_use(bool, bool) (mesh_base.C:196)                                                                                                                                                   
==25250==    by 0x582B0BA: libMesh::MeshTools::Generation::build_cube(libMesh::UnstructuredMesh&, unsigned int, unsigned int, unsigned int, double, double, double, double, double, double, libMesh::ElemType, bool) (mesh_generation.C:1455) 
==25250==    by 0x413AC9: main (vector_fe_ex3.C:88)                                                                                                                                                                                           
==25250==  Address 0x1 is not stack'd, malloc'd or (recently) free'd   
```
```
==25250== Unaddressable byte(s) found during client check request                                                                                                                                                                             
==25250==    at 0x4E415E0: check_mem_is_defined_untyped (libmpiwrap.c:953)                                                                                                                                                                    
==25250==    by 0x4E421DA: walk_type (libmpiwrap.c:691)                                                                                                                                                                                       
==25250==    by 0x4E49CE5: walk_type_array (libmpiwrap.c:924)                                                                                                                                                                                 
==25250==    by 0x4E49CE5: check_mem_is_defined (libmpiwrap.c:995)                                                                                                                                                                            
==25250==    by 0x4E49CE5: PMPI_Allreduce (libmpiwrap.c:2094)                                                                                                                                                                                 
==25250==    by 0x599FCCA: sum<unsigned int, std::allocator<unsigned int> > (parallel_implementation.h:753)                                                                                                                                   
==25250==    by 0x599FCCA: libMesh::Parallel::Histogram<Hilbert::HilbertIndices, unsigned int>::build_histogram() (parallel_histogram.C:101)                                                                                                  
==25250==    by 0x599A13F: libMesh::Parallel::BinSorter<Hilbert::HilbertIndices, unsigned int>::binsort(unsigned int, Hilbert::HilbertIndices, Hilbert::HilbertIndices) (parallel_bin_sorter.C:65)                                            
==25250==    by 0x59A0F27: libMesh::Parallel::Sort<Hilbert::HilbertIndices, unsigned int>::binsort() (parallel_sort.C:184)                                                                                                                    
==25250==    by 0x59AB809: libMesh::Parallel::Sort<Hilbert::HilbertIndices, unsigned int>::sort() (parallel_sort.C:82)                                                                                                                        
==25250==    by 0x581ED87: void libMesh::MeshCommunication::find_global_indices<libMesh::MeshBase::element_iterator>(libMesh::Parallel::Communicator const&, libMesh::BoundingBox const&, libMesh::MeshBase::element_iterator const&, libMesh\
::MeshBase::element_iterator const&, std::vector<unsigned int, std::allocator<unsigned int> >&) const (mesh_communication_global_indices.C:761)                                                                                              
==25250==    by 0x59C289F: libMesh::Partitioner::partition_unpartitioned_elements(libMesh::MeshBase&, unsigned int) (partitioner.C:241)                                                                                                       
==25250==    by 0x59C9AFD: libMesh::Partitioner::partition(libMesh::MeshBase&, unsigned int) (partitioner.C:83)                                                                                                                               
==25250==    by 0x57E618B: partition (mesh_base.h:750)                                                                                                                                                                                        
==25250==    by 0x57E618B: libMesh::MeshBase::prepare_for_use(bool, bool) (mesh_base.C:242)                                                                                                                                                   
==25250==    by 0x582B0BA: libMesh::MeshTools::Generation::build_cube(libMesh::UnstructuredMesh&, unsigned int, unsigned int, unsigned int, double, double, double, double, double, double, libMesh::ElemType, bool) (mesh_generation.C:1455) 
==25250==  Address 0x1 is not stack'd, malloc'd or (recently) free'd 
```
I also get complaints regarding leaked memory:
```
==25250== 560 bytes in 2 blocks are definitely lost in loss record 31 of 38                                                                                                                                                                 
==25250==    at 0x4C2CABB: malloc (vg_replace_malloc.c:299)                                                                                                                                                                                 
==25250==    by 0x9882CCF: MPID_Dataloop_struct_alloc (in /ssd/opt/moose/mpich-3.2/gcc-7.3.0/lib/libmpi.so.12.1.0)                                                                                                                          
==25250==    by 0x9888460: MPID_Dataloop_create_struct (in /ssd/opt/moose/mpich-3.2/gcc-7.3.0/lib/libmpi.so.12.1.0)                                                                                                                         
==25250==    by 0x9884AE4: MPID_Dataloop_create (in /ssd/opt/moose/mpich-3.2/gcc-7.3.0/lib/libmpi.so.12.1.0)                                                                                                                                
==25250==    by 0x989EDC4: MPID_Type_commit (in /ssd/opt/moose/mpich-3.2/gcc-7.3.0/lib/libmpi.so.12.1.0)                                                                                                                                    
==25250==    by 0x972721C: MPIR_Type_commit_impl (in /ssd/opt/moose/mpich-3.2/gcc-7.3.0/lib/libmpi.so.12.1.0)                                                                                                                               
==25250==    by 0x97275A3: PMPI_Type_commit (in /ssd/opt/moose/mpich-3.2/gcc-7.3.0/lib/libmpi.so.12.1.0)                                                                                                                                    
==25250==    by 0x4E443B7: PMPI_Type_commit (libmpiwrap.c:1769)                                                                                                                                                                             
==25250==    by 0x58F1EE6: StandardType (standard_type.h:163)                                                                                                                                                                               
==25250==    by 0x58F1EE6: void libMesh::Parallel::push_parallel_vector_data<std::map<unsigned short, std::vector<std::pair<Hilbert::HilbertIndices, unsigned long>, std::allocator<std::pair<Hilbert::HilbertIndices, unsigned long> > >, \
std::less<unsigned short>, std::allocator<std::pair<unsigned short const, std::vector<std::pair<Hilbert::HilbertIndices, unsigned long>, std::allocator<std::pair<Hilbert::HilbertIndices, unsigned long> > > > > >, std::vector<libMesh::P\
arallel::Request, std::allocator<libMesh::Parallel::Request> >, libMesh::Parallel::Sort<std::pair<Hilbert::HilbertIndices, unsigned long>, unsigned int>::communicate_bins()::{lambda(unsigned short, std::vector<std::pair<Hilbert::Hilber\
tIndices, unsigned long>, std::allocator<std::pair<Hilbert::HilbertIndices, unsigned long> > > const&)#1}>(libMesh::Parallel::Communicator const&, std::map<unsigned short, std::vector<std::pair<Hilbert::HilbertIndices, unsigned long>, \
std::allocator<std::pair<Hilbert::HilbertIndices, unsigned long> > >, std::less<unsigned short>, std::allocator<std::pair<unsigned short const, std::vector<std::pair<Hilbert::HilbertIndices, unsigned long>, std::allocator<std::pair<Hil\
bert::HilbertIndices, unsigned long> > > > > > const&, std::vector<libMesh::Parallel::Request, std::allocator<libMesh::Parallel::Request> >&, libMesh::Parallel::Sort<std::pair<Hilbert::HilbertIndices, unsigned long>, unsigned int>::com\
municate_bins()::{lambda(unsigned short, std::vector<std::pair<Hilbert::HilbertIndices, unsigned long>, std::allocator<std::pair<Hilbert::HilbertIndices, unsigned long> > > const&)#1}&) (parallel_sync.h:287)                             
==25250==    by 0x58F37D0: push_parallel_vector_data<std::map<short unsigned int, std::vector<std::pair<Hilbert::HilbertIndices, long unsigned int>, std::allocator<std::pair<Hilbert::HilbertIndices, long unsigned int> > >, std::less<sh\
ort unsigned int>, std::allocator<std::pair<short unsigned int const, std::vector<std::pair<Hilbert::HilbertIndices, long unsigned int>, std::allocator<std::pair<Hilbert::HilbertIndices, long unsigned int> > > > > >, libMesh::Parallel:\
:Sort<KeyType, IdxType>::communicate_bins() [with KeyType = std::pair<Hilbert::HilbertIndices, long unsigned int>; IdxType = unsigned int]::<lambda(libMesh::processor_id_type, const std::vector<std::pair<Hilbert::HilbertIndices, long u\
nsigned int>, std::allocator<std::pair<Hilbert::HilbertIndices, long unsigned int> > >&)> > (parallel_sync.h:441)                                                                                                                           
==25250==    by 0x58F37D0: libMesh::Parallel::Sort<std::pair<Hilbert::HilbertIndices, unsigned long>, unsigned int>::communicate_bins() (parallel_sort.C:225)                                                                               
==25250==    by 0x58F5101: libMesh::Parallel::Sort<std::pair<Hilbert::HilbertIndices, unsigned long>, unsigned int>::sort() (parallel_sort.C:83)                                                                                            
==25250==    by 0x576F307: void libMesh::MeshCommunication::find_global_indices<libMesh::MeshBase::element_iterator>(libMesh::Parallel::Communicator const&, libMesh::BoundingBox const&, libMesh::MeshBase::element_iterator const&, libMe\
sh::MeshBase::element_iterator const&, std::vector<unsigned int, std::allocator<unsigned int> >&) const (mesh_communication_global_indices.C:761) 
:
:
:
==25250== LEAK SUMMARY:                                                                                                                                                                                                                     
==25250==    definitely lost: 3,696 bytes in 27 blocks                                                                                                                                                                                      
==25250==    indirectly lost: 0 bytes in 0 blocks                                                                                                                                                                                           
==25250==      possibly lost: 0 bytes in 0 blocks                                                                                                                                                                                           
==25250==    still reachable: 123,552 bytes in 17 blocks                                                                                                                                                                                    
==25250==         suppressed: 0 bytes in 0 blocks                                                                                                                                                                                           
==25250== Reachable blocks (those to which a pointer was found) are not shown.                                                                                                                                                              
==25250== To see them, rerun with: --leak-check=full --show-leak-kinds=all                                                                                                                                                                  
==25250==                                                                                                                                                                                                                                   
==25250== For counts of detected and suppressed errors, rerun with: -v             
```

I get *some* of the same "Unaddressable byte" traces if I run with my own built `openmpi`. `openmpi` also yields a much larger number of leaked bytes, but `openmpi` is well known to not be totally valgrind clean.

There are 31 unaddressable byte errors. 6 of them cite `Alltoall`. The other 25 are from `Allreduce`. I did run an example mpi `Allreduce`program with `mpich`, and valgrind came back clean. There is definitely something wrong here, whether it's with valgrind or libMesh...

**Comment by lindsayad** [comment]
I will note that when running with `example-dbg`, the number of "Unaddressable bytes" errors drops to 24 (2 from `Alltoall`, 22 from `Allreduce`. The leaked memory results are the same.

**Comment by roystgnr** [comment]
On Tue, 11 Sep 2018, Alex Lindsay wrote:

> It is possible that this is a red herring, but so far googles of "Unaddressable bytes mpiwrap false positives" have yielded no results. If I run
> examples/vector_fe/ex3 with LD_PRELOAD=/path/to/libmpiwrap mpirun -np 2 valgrind ./example-oprof where mpirun is provided by the MOOSE package
> mpich, then I get these types of traces:
> 
> ==25250== Unaddressable byte(s) found during client check request

I have no clue abou this, unless not all mpich builds are created
equal?

https://lists.mcs.anl.gov/pipermail/petsc-users/2012-January/011824.html

> I also get complaints regarding leaked memory:
> 
> ==25250== 560 bytes in 2 blocks are definitely lost in loss record 31 of 38
> 
> ==25250==    at 0x4C2CABB: malloc (vg_replace_malloc.c:299)

> ==25250==    by 0x58F1EE6: StandardType (standard_type.h:163)

> ndices, unsigned long> > > const&)#1}&) (parallel_sync.h:287)

In theory:

   // This object will be free'd before all non-blocking communications
   // complete, but the MPI standard for MPI_Type_free specifies "Any
   // communication that is currently using this datatype will
   // complete normally." so we're cool.

And for this code path those communications complete on parallel_sync.h:443

In practice... is it possible that MPICH isn't doing garbage collection in
MPI_Wait but is just letting user-defined-then-freed-while-in-use types leak?
---
Roy

**Comment by lindsayad** [comment]
> I have no clue abou this, unless not all mpich builds are created equal? https://lists.mcs.anl.gov/pipermail/petsc-users/2012-January/011824.html

I also ran across this exact thread. I know the PETSc team is huge on maintaining valgrind cleanliness. It looks like PETSc currently downloads `'http://www.mpich.org/static/downloads/3.3b1/mpich-3.3b1.tar.gz'`. We are currently downloading `http://www.mpich.org/static/downloads/3.2/mpich-3.2.tar.gz`, so same distributor...but I suppose there's a chance there are issues with specific versions? When I have some more time, I'm going to try the PETSc `mpich` version.

> In practice... is it possible that MPICH isn't doing garbage collection in MPI_Wait but is just letting user-defined-then-freed-while-in-use types leak?

I don't know the answer to this...it is worth some more digging because we would like to set-up a test target for parallel valgrind.

**Comment by roystgnr** [comment]
A parallel valgrind test would be awesome.  I know we used to be leaking MPI datatypes, and although that got fixed it's entirely possible that there's been a regression or that we've added a new leak elsewhere thanks to the current lack of valgrind testing of parallel code paths.

**Comment by lindsayad** [comment]
With mpich-3.3b1:

### oprof

- 30 unaddressable byte errors
    - 6 `Alltoall`
    - 24 `Allreduce`
- 584 bytes lost in 6 blocks

### dbg

- 24 unaddressable byte errors
    - 2 `Alltoall`
    - 22 `Allreduce`
- 584 bytes lost in 6 blocks

Looks like essentially nothing changes when bumping the mpich version

**Comment by lindsayad** [comment]
So, the verdict on the unaddressable bytes errors is that...they're not our fault. I get those unaddressable byte errors with an MPI only example:
```c++
#include <mpi.h>                                                                                                                                                                                                                            
#include <iostream>                                                                                                                                                                                                                         
                                                                                                                                                                                                                                            
int main(int /*argc*/, char** /*argv*/)                                                                                                                                                                                                     
{                                                                                                                                                                                                                                           
  MPI_Init(NULL, NULL);                                                                                                                                                                                                                     
                                                                                                                                                                                                                                            
  int world_rank;                                                                                                                                                                                                                           
  MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);                                                                                                                                                                                               
                                                                                                                                                                                                                                            
  int i = world_rank;                                                                                                                                                                                                                       
  MPI_Allreduce(MPI_IN_PLACE, &i, 1, MPI_INT, MPI_MAX, MPI_COMM_WORLD);                                                                                                                                                                     
                                                                                                                                                                                                                                            
  if (world_rank == 0)                                                                                                                                                                                                                      
    std::cout << i << std::endl;                                                                                                                                                                                                            
                                                                                                                                                                                                                                            
  MPI_Finalize();                                                                                                                                                                                                                           
                                                                                                                                                                                                                                            
  return 0;                                                                                                                                                                                                                                 
}
```
So, the culprit is a lack of valgrind support for `MPI_IN_PLACE`, as described in this [bug report](https://bugs.kde.org/show_bug.cgi?id=287862). It looks like there is a fix, but it hasn't been merged. It seems like we could create a suppressions file for all the `Communicator` routines that use `MPI_IN_PLACE`. Thoughts?

**Comment by roystgnr** [comment]
Thanks for pinning that down!  Let's *only* suppress the MPI_IN_PLACE "errors" until we understand where the remaining leaks are coming from, to be clear.  If 90% of the leaked bytes were fixed by an mpich upgrade then I wouldn't be surprised if the other 10% aren't our fault, but I'd like to understand them before acting confident about that.

**Comment by lindsayad** [comment]
Ok, I'm going to make a PR into MOOSE soon with suppressions for the unaddressable bytes.

For leaks, some appear to be coming from PETSc (don't know whether they're real or not...tag @fdkong ). Here's the leak summary from the `vector_fe ex3` with mpich3.3b:
```
==11559== 32 bytes in 1 blocks are definitely lost in loss record 4 of 24                                                                                                                                                                   
==11559==    at 0x4C2DBF6: malloc (vg_replace_malloc.c:299)                                                                                                                                                                                 
==11559==    by 0x735A5FF: MPIR_Type_contiguous_impl (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                             
==11559==    by 0x735AC3C: PMPI_Type_contiguous (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                                  
==11559==    by 0x4E65BB2: PMPI_Type_contiguous (libmpiwrap.c:2718)                                                                                                                                                                         
==11559==    by 0x599E5BB: DataType (data_type.h:81)                                                                                                                                                                                        
==11559==    by 0x599E5BB: StandardType (parallel_hilbert.h:60)                                                                                                                                                                             
==11559==    by 0x599E5BB: libMesh::Parallel::Sort<Hilbert::HilbertIndices, unsigned int>::binsort() (parallel_sort.C:170)                                                                                                                  
==11559==    by 0x59A8C18: libMesh::Parallel::Sort<Hilbert::HilbertIndices, unsigned int>::sort() (parallel_sort.C:82)                                                                                                                      
==11559==    by 0x581C347: void libMesh::MeshCommunication::find_global_indices<libMesh::MeshBase::element_iterator>(libMesh::Parallel::Communicator const&, libMesh::BoundingBox const&, libMesh::MeshBase::element_iterator const&, libMe\
sh::MeshBase::element_iterator const&, std::vector<unsigned int, std::allocator<unsigned int> >&) const (mesh_communication_global_indices.C:761)                                                                                           
==11559==    by 0x59BFCAF: libMesh::Partitioner::partition_unpartitioned_elements(libMesh::MeshBase&, unsigned int) (partitioner.C:241)                                                                                                     
==11559==    by 0x59C6EFD: libMesh::Partitioner::partition(libMesh::MeshBase&, unsigned int) (partitioner.C:83)                                                                                                                             
==11559==    by 0x57E3AAB: partition (mesh_base.h:750)                                                                                                                                                                                      
==11559==    by 0x57E3AAB: libMesh::MeshBase::prepare_for_use(bool, bool) (mesh_base.C:242)                                                                                                                                                 
==11559==    by 0x582867A: libMesh::MeshTools::Generation::build_cube(libMesh::UnstructuredMesh&, unsigned int, unsigned int, unsigned int, double, double, double, double, double, double, libMesh::ElemType, bool) (mesh_generation.C:145\
5)                                                                                                                                                                                                                                          
==11559==    by 0x413A59: main (vector_fe_ex3.C:88)                                                                                                                                                                                         
==11559==                                                                                                                                                                                                                                   
==11559== 72 bytes in 1 blocks are definitely lost in loss record 11 of 24                                                                                                                                                                  
==11559==    at 0x4C2DBF6: malloc (vg_replace_malloc.c:299)                                                                                                                                                                                 
==11559==    by 0x740216F: MPIR_Dataloop_alloc_and_copy (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                          
==11559==    by 0x7404BAB: MPIR_Dataloop_create_contiguous (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                       
==11559==    by 0x7402BF0: MPIR_Dataloop_create (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                                  
==11559==    by 0x7355FFF: MPIR_Type_commit (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                                      
==11559==    by 0x73560EE: MPIR_Type_commit_impl (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                                 
==11559==    by 0x7356338: PMPI_Type_commit (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                                      
==11559==    by 0x4E468F0: PMPI_Type_commit (libmpiwrap.c:1769)                                                                                                                                                                             
==11559==    by 0x599E5C3: commit (data_type.h:111)                                                                                                                                                                                         
==11559==    by 0x599E5C3: DataType (data_type.h:82)                                                                                                                                                                                        
==11559==    by 0x599E5C3: StandardType (parallel_hilbert.h:60)                                                                                                                                                                             
==11559==    by 0x599E5C3: libMesh::Parallel::Sort<Hilbert::HilbertIndices, unsigned int>::binsort() (parallel_sort.C:170)                                                                                                                  
==11559==    by 0x59A8C18: libMesh::Parallel::Sort<Hilbert::HilbertIndices, unsigned int>::sort() (parallel_sort.C:82)                                                                                                                      
==11559==    by 0x581C347: void libMesh::MeshCommunication::find_global_indices<libMesh::MeshBase::element_iterator>(libMesh::Parallel::Communicator const&, libMesh::BoundingBox const&, libMesh::MeshBase::element_iterator const&, libMe\
sh::MeshBase::element_iterator const&, std::vector<unsigned int, std::allocator<unsigned int> >&) const (mesh_communication_global_indices.C:761)                                                                                           
==11559==    by 0x59BFCAF: libMesh::Partitioner::partition_unpartitioned_elements(libMesh::MeshBase&, unsigned int) (partitioner.C:241)                                                                                                     
==11559==                                                                                                                                                                                                                                   
==11559== 72 bytes in 1 blocks are definitely lost in loss record 12 of 24                                                                                                                                                                  
==11559==    at 0x4C2DBF6: malloc (vg_replace_malloc.c:299)                                                                                                                                                                                 
==11559==    by 0x740216F: MPIR_Dataloop_alloc_and_copy (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                          
==11559==    by 0x7404BAB: MPIR_Dataloop_create_contiguous (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                       
==11559==    by 0x7402BF0: MPIR_Dataloop_create (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                                  
==11559==    by 0x7356022: MPIR_Type_commit (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                                      
==11559==    by 0x73560EE: MPIR_Type_commit_impl (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                                 
==11559==    by 0x7356338: PMPI_Type_commit (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                                      
==11559==    by 0x4E468F0: PMPI_Type_commit (libmpiwrap.c:1769)                                                                                                                                                                             
==11559==    by 0x599E5C3: commit (data_type.h:111)                                                                                                                                                                                         
==11559==    by 0x599E5C3: DataType (data_type.h:82)                                                                                                                                                                                        
==11559==    by 0x599E5C3: StandardType (parallel_hilbert.h:60)                                                                                                                                                                             
==11559==    by 0x599E5C3: libMesh::Parallel::Sort<Hilbert::HilbertIndices, unsigned int>::binsort() (parallel_sort.C:170)                                                                                                                  
==11559==    by 0x59A8C18: libMesh::Parallel::Sort<Hilbert::HilbertIndices, unsigned int>::sort() (parallel_sort.C:82)                                                                                                                      
==11559==    by 0x581C347: void libMesh::MeshCommunication::find_global_indices<libMesh::MeshBase::element_iterator>(libMesh::Parallel::Communicator const&, libMesh::BoundingBox const&, libMesh::MeshBase::element_iterator const&, libMe\
sh::MeshBase::element_iterator const&, std::vector<unsigned int, std::allocator<unsigned int> >&) const (mesh_communication_global_indices.C:761)                                                                                           
==11559==    by 0x59BFCAF: libMesh::Partitioner::partition_unpartitioned_elements(libMesh::MeshBase&, unsigned int) (partitioner.C:241)                                                                                                     
==11559==                                                                                                                                                                                                                                   
==11559== 112 bytes in 2 blocks are definitely lost in loss record 14 of 24                                                                                                                                                                 
==11559==    at 0x4C2DBF6: malloc (vg_replace_malloc.c:299)                                                                                                                                                                                 
==11559==    by 0x7361237: MPIR_Type_create_struct_impl (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                          
==11559==    by 0x73619B6: PMPI_Type_create_struct (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                               
==11559==    by 0x4E66FC5: PMPI_Type_create_struct (libmpiwrap.c:2727)                                                                                                                                                                      
==11559==    by 0x82A21AD: MatStashBlockTypeSetUp (matstash.c:785)                                                                                                                                                                          
==11559==    by 0x82A21AD: MatStashScatterBegin_BTS (matstash.c:852)                                                                                                                                                                        
==11559==    by 0x82A51F5: MatStashScatterBegin_Private (matstash.c:462)                                                                                                                                                                    
==11559==    by 0x823161B: MatAssemblyBegin_MPIAIJ (mpiaij.c:774)                                                                                                                                                                           
==11559==    by 0x7FD800B: MatAssemblyBegin (matrix.c:5247)                                                                                                                                                                                 
==11559==    by 0x5981C44: libMesh::PetscMatrix<double>::close() (petsc_matrix.C:963)                                                                                                                                                       
==11559==    by 0x5AEB484: libMesh::PetscLinearSolver<double>::solve(libMesh::SparseMatrix<double>&, libMesh::SparseMatrix<double>&, libMesh::NumericVector<double>&, libMesh::NumericVector<double>&, double, unsigned int) (petsc_linear_\
solver.C:455)                                                                                                                                                                                                                               
==11559==    by 0x5AD1985: libMesh::PetscLinearSolver<double>::solve(libMesh::SparseMatrix<double>&, libMesh::NumericVector<double>&, libMesh::NumericVector<double>&, double, unsigned int) (petsc_linear_solver.h:163)                    
==11559==    by 0x5AD8BE2: solve (linear_solver.h:340)                                                                                                                                                                                      
==11559==    by 0x5AD8BE2: libMesh::NewtonSolver::solve() (newton_solver.C:405)                                                                                                                                                             
==11559==                                                                                                                                                                                                                                   
==11559== 144 bytes in 2 blocks are definitely lost in loss record 15 of 24                                                                                                                                                                 
==11559==    at 0x4C2DBF6: malloc (vg_replace_malloc.c:299)                                                                                                                                                                                 
==11559==    by 0x740216F: MPIR_Dataloop_alloc_and_copy (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                          
==11559==    by 0x7404BAB: MPIR_Dataloop_create_contiguous (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                       
==11559==    by 0x74055B2: MPIR_Dataloop_create_indexed (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                          
==11559==    by 0x7406C5A: MPIR_Dataloop_create_struct (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                           
==11559==    by 0x74036BD: MPIR_Dataloop_create (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                                  
==11559==    by 0x7355FFF: MPIR_Type_commit (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                                      
==11559==    by 0x73560EE: MPIR_Type_commit_impl (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                                 
==11559==    by 0x7356338: PMPI_Type_commit (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                                      
==11559==    by 0x4E468F0: PMPI_Type_commit (libmpiwrap.c:1769)                                                                                                                                                                             
==11559==    by 0x82A21FB: MatStashBlockTypeSetUp (matstash.c:786)                                                                                                                                                                          
==11559==    by 0x82A21FB: MatStashScatterBegin_BTS (matstash.c:852)                                                                                                                                                                        
==11559==    by 0x82A51F5: MatStashScatterBegin_Private (matstash.c:462)                                                                                                                                                                    
==11559==                                                                                                                                                                                                                                   
==11559== 560 bytes in 2 blocks are definitely lost in loss record 17 of 24                                                                                                                                                                 
==11559==    at 0x4C2DBF6: malloc (vg_replace_malloc.c:299)                                                                                                                                                                                 
==11559==    by 0x7402414: MPIR_Dataloop_struct_alloc (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                            
==11559==    by 0x74064AA: MPIR_Dataloop_create_struct (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                           
==11559==    by 0x74036BD: MPIR_Dataloop_create (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                                  
==11559==    by 0x7356022: MPIR_Type_commit (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                                      
==11559==    by 0x73560EE: MPIR_Type_commit_impl (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                                 
==11559==    by 0x7356338: PMPI_Type_commit (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                                      
==11559==    by 0x4E468F0: PMPI_Type_commit (libmpiwrap.c:1769)                                                                                                                                                                             
==11559==    by 0x82A21FB: MatStashBlockTypeSetUp (matstash.c:786)                                                                                                                                                                          
==11559==    by 0x82A21FB: MatStashScatterBegin_BTS (matstash.c:852)                                                                                                                                                                        
==11559==    by 0x82A51F5: MatStashScatterBegin_Private (matstash.c:462)                                                                                                                                                                    
==11559==    by 0x823161B: MatAssemblyBegin_MPIAIJ (mpiaij.c:774)                                                                                                                                                                           
==11559==    by 0x7FD800B: MatAssemblyBegin (matrix.c:5247)                                                                                                                                                                                 
==11559==                                                                                                                                                                                                                                   
==11559== LEAK SUMMARY:                                                                                                                                                                                                                     
==11559==    definitely lost: 992 bytes in 9 blocks                                                                                                                                                                                         
==11559==    indirectly lost: 0 bytes in 0 blocks                                                                                                                                                                                           
==11559==      possibly lost: 0 bytes in 0 blocks                                                                                                                                                                                           
==11559==    still reachable: 196,280 bytes in 18 blocks                                                                                                                                                                                    
==11559==         suppressed: 0 bytes in 0 blocks                                                                                                                                                                                           
==11559== Reachable blocks (those to which a pointer was found) are not shown.                                                                                                                                                              
==11559== To see them, rerun with: --leak-check=full --show-leak-kinds=all                                                                                                                                                                  
==11559==                                                                                                                                                                                                                                   
==11559== For counts of detected and suppressed errors, rerun with: -v                                                                                                                                                                      
==11559== ERROR SUMMARY: 6 errors from 6 contexts (suppressed: 250 from 30) 
```

As far as trying to understand these...I was looking in the constructor for `StandardType<HilbertIndices>`:
```c++
  StandardType(const Hilbert::HilbertIndices * =nullptr) {                                                                                                                                                                                  
    // _static_type never gets freed, but it only gets committed once                                                                                                                                                                       
    // so it's not a *huge* memory leak...                                                                                                                                                                                                  
    static DataType _static_type;                           
```
and I guess I don't understand the comment there. Aren't we just creating a static integer handle here? Shouldn't that be handled at the end of the program?

**Comment by roystgnr** [comment]
> I guess I don't understand the comment there.

Presumably you're starting from the conditional statement "A smart person wouldn't leave a memory leak lying around like that", and are having trouble applying modus ponens when what you really need is modus tollens.

> Aren't we just creating a static integer handle here? Shouldn't that be handled at the end of the program?

We're creating a handle, but then we're using MPI_Type_commit in the StandardType constructor and we're never doing a corresponding MPI_Type_free.  I fixed the same damn bug in #1616 but somehow I missed this instance.

**Comment by lindsayad** [comment]
Closed by #1860
